### PR TITLE
Feature/nextcloud

### PR DESCRIPTION
**Pull request by DutcherNL**
Developing source code to synch Nextcloud with Django.

To do:

- [x] Permissions to download
- [x] Permissions to edit
- [ ] Uploading to Nextcloud
- [x] General download page linked in header

Possible extentions:
- Connect NCFolder to activity (e.g. ALV including the files for that ALV)

Note: Login info is not included in the branch but can be set in your `squire/local_settings` file. Running Django automatically creates that file. `local_settings.py` is not synced with Git.

### CODE REVIEW

File: nextcloud_integration/templates/nextcloud_integration/snippets/nc_folder_contents.html [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        <h4 class="mb-0">
+            {{ folder.display_name }}
+        </h4>
+        <p class="mb-0">
+            {{ folder.description }}
+        </p>
+        {% if folder.is_missing %}
+            <div class="text-danger">
+                    <i class="fas fa-exclamation-triangle"></i>
+                    Folder missing on disk
```
Comment (EricTRL) [inline nextcloud_integration/templates/nextcloud_integration/snippets/nc_folder_contents.html:14]:
```suggestion
                    Folder missing
```

File: nextcloud_integration/templates/nextcloud_integration/snippets/nc_folder_contents.html [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                </div>
+            </li>
+        {% endfor %}
+        <li class="list-group-item">
+            {% if user|has_edit_access:folder %}
+                <a class="btn btn-primary" href="{% url "nextcloud:folder_edit" folder_slug=folder.slug %}">
+                    <i class="fas fa-pen"></i> Edit
+                </a>
+            {% endif %}
+            {% if user|has_synch_access %}
```
Comment (EricTRL) [inline nextcloud_integration/templates/nextcloud_integration/snippets/nc_folder_contents.html:61]:
```suggestion
            {% if perms.nextcloud_integration.synch_ncfile %}
```

File: nextcloud_integration/widgets.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from django.forms.widgets import ChoiceWidget
+
+
+class NextcloudFileSelectWidget(ChoiceWidget):
+    template_name = "nextcloud_integration/widgets/nextcloud_file_select_widget.html"
+    pass
+
+    # def render(self, name, value, attrs=None, renderer=None):
```
Comment (EricTRL) [inline nextcloud_integration/widgets.py:8]: :)

File: nextcloud_integration/views.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    template_name = "nextcloud_integration/site_downloads.html"
+    model = NCFolder
+    context_object_name = "folders"
+
+    def get_queryset(self):
+        queryset = super(SiteDownloadView, self).get_queryset()
+        queryset = queryset.filter(on_overview_page=True)
+
+        if not user_is_current_member(self.request.user):
+            queryset = queryset.filter(requires_membership=False)
```
[the remark this answers is not in the document]
Comment (DutcherNL) [reply nextcloud_integration/views.py:91]: I will move some logic, including this mechanic to a groupconfig later on

File: nextcloud_integration/models.py [hunk 5]
```diff
@@ -51,43 +51,53 @@ def __str__(self):
...
-        return super(NCFile, self).save(**kwargs)
+        return super(SquireNextCloudFile, self).save(**kwargs)
 
     def clean(self):
+        self._set_slug()
         if self.file is None and (self.file_name is None or self.file_name == ""):
-            raise ValidationError("Either file or filename should be defined")
+            raise ValidationError("Either file or filename should be defined", code="no_file_name_linked")
+
+        if SquireNextCloudFile.objects.filter(folder=self.folder, slug=self.slug).exclude(id=self.id).exists():
```
Comment (EricTRL) [inline nextcloud_integration/models.py:95]: Fine as is, but this is essentially a [unique_together](https://docs.djangoproject.com/en/3.2/ref/models/options/#unique-together) on the slug and folder

File: nextcloud_integration/forms.py [hunk 6]
```diff
@@ -60,27 +63,31 @@ def execute(self):
...
+    def __init__(self, *args, folder: SquireNextCloudFolder=None, **kwargs):
         assert folder is not None
         self.folder = folder
-        self.file_list = self.get_unsynched_files()
+        self.file_list = self.get_unsynced_files()
 
-        super(SynchFileToFolderForm, self).__init__(*args, **kwargs)
+        super(SyncFileToFolderForm, self).__init__(*args, **kwargs)
         self.instance.folder = self.folder
         self.instance.connection = "NcS"
```
Comment (EricTRL) [inline nextcloud_integration/forms.py:92]: Might as well use the new value :)
```suggestion
        self.instance.connection = SquireNextCloudFile.CONNECTION_NEXTCLOUD_SYNC
```

File: nextcloud_integration/migrations/0001_initial.py [hunk 7]
```diff
@@ -1,34 +1,46 @@
-# Generated by Django 2.2.25 on 2022-09-22 21:08
+# Generated by Django 3.2.16 on 2023-01-08 16:00
 
 from django.db import migrations, models
 import django.db.models.deletion
 
 
 class Migration(migrations.Migration):
 
+    replaces = [('nextcloud_integration', '0001_initial'), ('nextcloud_integration', '0002_auto_20220925_1758'), ('nextcloud_integration', '0003_ncfolder_slug'), ('nextcloud_integration', '0004_auto_20221001_1612'), ('nextcloud_integration', '0005_auto_20221012_1618'), ('nextcloud_integration', '0006_auto_20221016_1845'), ('nextcloud_integration', '0007_auto_20221107_2228'), ('nextcloud_integration', '0008_auto_20230108_1558'), ('nextcloud_integration', '0009_auto_20230108_1559'), ('nextcloud_integration', '0010_auto_20230108_1642')]
```
Comment (EricTRL) [inline nextcloud_integration/migrations/0001_initial.py]: `replaces` also needs to be removed when the squashed-migration files themselves are also removed. See the last bullet point at the bottom of [this](https://docs.djangoproject.com/en/3.2/topics/migrations/#squashing-migrations) docs section.

>You must then transition the squashed migration to a normal migration by:
> - Deleting all the migration files it replaces.
>- Updating all migrations that depend on the deleted migrations to depend on the squashed migration instead.
>- Removing the **replaces** attribute in the **Migration** class of the squashed migration (this is how Django tells that it is a squashed migration).

**Review by EricTRL — CHANGES_REQUESTED** [review]

**Review by EricTRL — APPROVED** [review]

**Pull request merged**